File: JUNGOL/Beginner/#5398 달팽이트리/달팽이트리.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
	int n; cin>>n;
	vector<vector<int>> snail;
	for(int i=0; i<n; i++){
		vector<int> v(n);
		snail.push_back(v);
	}

	int x=-1, y=0, num=0;
	int cnt=n, dir=0;	// 아래, 오른쪽, 대각선
	while(cnt){
		if(dir == 0){
			for(int i=0; i<cnt; i++){
				snail[++x][y] = num++;
				if(num >= 10) num = 0;
			}
		}
		else if(dir == 1){
			for(int i=0; i<cnt; i++){
				snail[x][++y] = num++;
				if(num >= 10) num = 0;
			}
		}
		else{
			for(int i=0; i<cnt; i++){
				snail[--x][--y] = num++;
				if(num >= 10) num = 0;
			}
		}

		dir = (dir+1)%3;
		cnt--;
	}

	for(int i=0; i<n; i++){
		for(int j=0; j<n-i-1; j++) printf(" ");
		for(int j=0; j<=i; j++) printf("%d ", snail[i][j]);
		cout<<endl;
	}
}

File: JUNGOL/Beginner/#5398 달팽이트리/달팽이트리.py
def main():
    n = int(input())
    snail = [[0 for i in range(n)] for i in range(n)]

    cnt,dir = n,0
    x,y,num = -1,0,0
    while cnt > 0:
        if dir == 0:
            for i in range(cnt):
                x += 1
                snail[x][y] = num
                num = (num+1)%10
        elif dir == 1:
            for i in range(cnt):
                y += 1
                snail[x][y] = num
                num = (num+1)%10
        else:
            for i in range(cnt):
                x,y = x-1,y-1
                snail[x][y] = num
                num = (num+1)%10

        dir = (dir+1)%3
        cnt -= 1

    for i in range(n):
        print(' '*(n-i-1), end='')
        for j in range(i+1): print(snail[i][j], end=' ')
        print()

main()
